bytecomp/matching.ml

```diff
@@ -1178,160 +1193,66 @@ let rec split_or argo cls args def =
           (Default_environment.cons matrix idef def, (idef, next) :: nexts)
     in
     match yesor with
-    | [] -> split_constr yes args def nexts
+    | [] -> split_no_or yes args def nexts
     | _ -> precompile_or argo yes yesor args def nexts
   in
   do_split [] [] [] cls
 
-(* Ultra-naive splitting, close to semantics, used for extension,
-   as potential rebind prevents any kind of optimisation
-
-   Indeed, extension constructors with distinct names may be equal thanks to
-   constructor rebinding. This is compiled by having a specialized
-   submatrix for each syntactically-distinct constructor, with a threading
-   of exits such that each submatrix falls back to the potentially-compatible
-   submatrices below it.
-*)
-and split_naive cls args def k =
-  let rec split_exc cstr0 rev_yes = function
-    | [] ->
-        let yes = List.rev rev_yes in
-        ( { me = Pm { cases = yes; args; default = def };
-            matrix = as_matrix yes;
-            top_default = def
-          },
-          k )
-    | ((p :: _, _) as cl) :: rem ->
-        if group_constructor p then
-          let cstr = pat_as_constr p in
-          if cstr = cstr0 then
-            split_exc cstr0 (cl :: rev_yes) rem
-          else
-            let yes = List.rev rev_yes in
-            let { me = next; matrix; top_default = def }, nexts =
-              split_exc cstr [ cl ] rem
-            in
-            let idef = next_raise_count () in
-            let def = Default_environment.cons matrix idef def in
-            ( { me = Pm { cases = yes; args; default = def };
-                matrix = as_matrix yes;
-                top_default = def
-              },
-              (idef, next) :: nexts )
-        else
-          let yes = List.rev rev_yes in
-          let { me = next; matrix; top_default = def }, nexts =
-            split_noexc [ cl ] rem
-          in
-          let idef = next_raise_count () in
-          let def = Default_environment.cons matrix idef def in
-          ( { me = Pm { cases = yes; args; default = def };
-              matrix = as_matrix yes;
-              top_default = def
-            },
-            (idef, next) :: nexts )
-    | _ -> assert false
-  and split_noexc rev_yes = function
+and split_no_or cls args def k =
+  let rec split cls =
+    let discr = what_is_first_case cls in
+    collect discr [] [] cls
+  and collect group_discr rev_yes rev_no = function
     | ([], _) :: _ -> assert false
-    | [] -> precompile_var args (List.rev rev_yes) def k
-    | ((p :: _, _) as cl) :: rem ->
-        if not (group_constructor p) then
-          split_noexc (cl :: rev_yes) rem
+    | [ ((ps, _) as cl) ] when List.for_all group_var ps && rev_yes <> [] ->
+        (* This enables an extra division in some frequent cases:
+               last row is made of variables only *)
+        collect group_discr rev_yes (cl :: rev_no) []
```

Comment:
I maybe have an idea of what this optimization is good for (as far as I know, we still don't understand this point). If we did *not* split here, then the whole group would have this catch-all row in its default-environment matrix, and it would never disappear due to further specialize. With the split here, then the first part of the group may be specialized away at some point, and the one-row group that we are creating would remain alone in the default environment; pattern matchings that fail earlier on would jump to it directly, which is more efficient than going through the first part of the matrix first.

Reply:
Note: the patterns in the tail of `ps` are not in simple form, so `group_var` is not the right function to call here -- in particular, it does not accept `Tpat_var` which is likely to occur. It would make sense to call `fun p -> le_pat p omega` -- which is also constant-time.

Reply:
> maybe have an idea of what this optimization is good for ...

Yes, I had reached the same conclusion the other day. It is in fact fairly easy to write a test case demonstrating it:
```ocaml
let f = function
  | true, false -> 1
  | _, false -> 2
  | _, _ -> 3
```
On trunk, 4.08, and this branch the output of `ocamlc -dlambda -c -nostdlib -nopervasives ./test.ml` is:
```
(setglobal Test!
  (let
    (f/3 =
       (function param/4 : int
         (catch
           (if (!= (field 0 param/4) 0)
             (if (!= (field 1 param/4) 0) (exit 1) 1)
             (if (!= (field 1 param/4) 0) (exit 1) 2))
          with (1) 3)))
    (makeblock 0 f/3)))
```
Now, if I comment out the optimisation clause, it changes to the following:
```
(setglobal Test!
  (let
    (f/3 =
       (function param/4 : int
         (catch
           (if (!= (field 0 param/4) 0)
             (if (!= (field 1 param/4) 0) (exit 1) 1) (exit 1))
          with (1) (if (!= (field 1 param/4) 0) 3 2))))
    (makeblock 0 f/3)))
```
Should I add this to the testsuite and add a comment in the code referring to it?

> Note: the patterns in the tail of ps are not in simple form, so group_var is not the right function to call here -- in particular, it does not accept Tpat_var which is likely to occur. It would make sense to call fun p -> le_pat p omega -- which is also constant-time.

Good catch indeed.
However, I can't help but notice that this is already the case in the original version, could it be on purpose?

Reply:
> It would make sense to call `fun p -> le_pat p omega` -- which is also constant-time.

I think that's wrong actually, let's say `p` is `C (D (E _))`. `le_pat p omega` ends up calling `not (satisfiable [[p]] [omega])`, and `satisfiable` will end up recursing a a few times (doing a bunch of non-`O(1)` operations each time).

I vote for not touching this for now.